lib/SILOptimizer/Mandatory/MandatoryInlining.cpp

```diff
@@ -996,40 +995,10 @@ class MandatoryInlining : public SILModuleTransform {
       if (SILVerifyAll) {
         F.verify();
       }
-    }
-
-    if (!ShouldCleanup)
-      return;
-
-    // Now that we've inlined some functions, clean up.  If there are any
-    // transparent functions that are deserialized from another module that are
-    // now unused, just remove them from the module.
-    //
-    // We do this with a simple linear scan, because transparent functions that
-    // reference each other have already been flattened.
-    for (auto FI = M->begin(), E = M->end(); FI != E; ) {
-      SILFunction &F = *FI++;
-
-      invalidateAnalysis(&F, SILAnalysis::InvalidationKind::Everything);
-
-      if (F.getRefCount() != 0) continue;
 
-      // Leave non-transparent functions alone.
-      if (!F.isTransparent())
-        continue;
-
-      // We discard functions that don't have external linkage,
-      // e.g. deserialized functions, internal functions, and thunks.
-      // Being marked transparent controls this.
-      if (F.isPossiblyUsedExternally()) continue;
-
-      // ObjC functions are called through the runtime and are therefore alive
-      // even if not referenced inside SIL.
-      if (F.getRepresentation() == SILFunctionTypeRepresentation::ObjCMethod)
-        continue;
-
-      // Okay, just erase the function from the module.
-      FuncBuilder.eraseFunction(&F);
+      if (!getOptions().DebugSerialization) {
+        invalidateAnalysis(&F, SILAnalysis::InvalidationKind::Everything);
```

Comment:
This is not sufficient without the pass manager change I mentioned. The inlining here is done recursively so you are going to need to also invalidate functions that are inlined into.

Otherwise, we will hit this non-deterministically. I think you can probably create a sil test case that does this. Let me describe it. If I am right, can you add a test for this?

Consider:

```
sil [transparent] @f {
  call g()
}
sil h() {
  doSomethingThatEffectsDominance()
}
sil [transparent] g() {
  h()
}
```

I think the optimizer will visit this here. The key thing you need to do is run a pass that constructs Dom Trees. I would be fine with an addition of a new optimizer pass that just does that.

Reply:
Hmm, good catch. Even though this pass runs on each function recursively, won't all the functions touched be in the module, so they will eventually, unconditionally (except if without `DebugSerialization`), get invalidated? 

I'm trying to work out a test case but, I'm not sure I understand what you mean by `doSomethingThatEffectsDominance`.
<details>
<summary>Here's what I've got so far (not much more than what you posted)</summary>

```
sil @f : $@convention(thin) () -> () {
bb0:
  // function_ref g()
  %0 = function_ref @g : $@convention(thin) () -> ()
  %1 = apply %0() : $@convention(thin) () -> ()
  %2 = tuple ()
  return %2 : $()
}

sil @h : $@convention(thin) () -> Builtin.Int1 {
bb0:
  %0 = integer_literal $Builtin.Int1, 0
  br bb2
bb1:
  return %0 : $Builtin.Int1
bb2:
  br bb1
}

sil [transparent] @g : $@convention(thin) () -> () {
bb0:
  // function_ref h()
  %0 = function_ref @h : $@convention(thin) () -> Builtin.Int1
  %1 = apply %0() : $@convention(thin) () -> Builtin.Int1
  %2 = tuple ()
  return %2 : $()
}
```

</details>

Reply:
I added a validation test to this PR, let me know what you think.